Support searching traces by span name

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17326?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17326/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17326/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17326/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Support searching traces by span name (undocumented).
*** Docs will be added once the MLflow client is logging spans into SQL DB by default.**

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [X] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
Add support for span name search functionality

- Add _SPAN_IDENTIFIER to SearchTraceUtils with support for span.name, span.type, span.status, and span.content
- Add is_span validation method with proper comparator validation
- Add span value parsing and comparison logic
- Update SqlAlchemy store to handle span filtering with subqueries
- Support span.name LIKE '%database%' searches in trace filtering

dbczumar committed:
Simplify span search to ONLY support span.name

- Restrict is_span() to only validate span.name attribute
- Simplify SqlAlchemy filtering to only handle span.name searches
- Remove support for span.type, span.status, and span.content for now
- Focus implementation on minimal span.name LIKE '%database%' functionality

dbczumar committed:
Remove span.name specific comments for future extension

- Remove 'Only support span.name for now' comments
- Remove 'Handle span.name filtering only' comments
- Remove TODO comment from span filtering
- Clean up code for future extension to other span attributes

dbczumar committed:
Remove hacky hasattr check for span subqueries

- Label span trace_id column as 'request_id' in subquery
- Remove hasattr check since all subqueries now have consistent column names
- Simplify join logic for non-attribute filters

dbczumar committed:
Remove unwanted R install deps file

- Remove mlflow/R/mlflow/source(".install-deps.R", echo = TRUE)
- Clean up repository by removing unintended file

dbczumar committed:
Add inline comment and exception handling for span filtering

- Explain why span filtering needs continue (direct columns vs key-value structure)
- Add exception for unsupported span attributes in database filtering
- Ensure proper error handling when span attribute isn't 'name'

dbczumar committed:
Remove redundant span attribute validation

- Remove if key_name == 'name' check since is_span() already validates this
- Remove redundant exception since validation happens upstream
- Simplify code by trusting upstream validation

dbczumar committed:
Add span name search test cases to test_sqlalchemy_store

- Remove perf_test.py
- Add test_search_traces_with_span_name_filter() with exact match, LIKE patterns, and NOT EQUAL
- Add test_search_traces_with_invalid_span_attribute() to test error handling
- Use synchronous log_spans method for simplicity

dbczumar committed:
Make span search more generic and add constants

- Add _SUPPORTED_SPAN_ATTRIBUTES constant to centralize supported span attributes
- Update is_span() to use constant instead of hardcoding 'name'
- Make SqlAlchemy store use getattr() for dynamic column access
- Prepare for easy extension to support more span attributes (type, status, etc.)
- Update error message to dynamically list supported attributes

dbczumar committed:
Fix: Create span-specific comparators to avoid affecting other attributes

- Revert VALID_STRING_ATTRIBUTE_COMPARATORS to original values (remove LIKE/ILIKE)
- Add VALID_SPAN_COMPARATORS with LIKE/ILIKE support specifically for spans
- Update is_span() to use span-specific comparators
- Ensure LIKE/ILIKE only available for spans, not all string attributes

dbczumar committed:
Rename VALID_SPAN_COMPARATORS to VALID_SPAN_ATTRIBUTE_COMPARATORS

- Rename constant to match naming pattern with other comparator constants
- Update references in is_span() method
- Maintain consistency with VALID_STRING_ATTRIBUTE_COMPARATORS naming

dbczumar committed:
Add test case that specifically matches trace2 in span name filter test

- Add test for exact match on 'api_call' to ensure trace2 is matched
- Verify single trace result and correct trace_id
- Improve test coverage for individual trace matching

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy reviewed (approved):
LGTM once lint errors are fixed!

dbczumar committed:
Fix span creation in test using create_test_span helper

- Replace incorrect create_mlflow_span() usage with create_test_span() helper
- Add experiment_id parameter to log_spans() calls as required
- Use proper span_id integers (111, 222, 333) instead of strings
- Fix test to match actual function signatures

dbczumar committed:
Fix line length issues in span search tests

- Break long lines in create_test_span() calls across multiple lines
- Break long store.search_traces() calls to keep under 100 chars
- Break long pytest.raises() calls for better readability
- All lines now comply with line length requirements

dbczumar committed:
Fix test and remove debug code

- Remove debugging code from test and sqlalchemy store
- Fix test to properly assign traces variable for first assertion
- Clean up code while keeping span search functionality

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge remote-tracking branch 'origin/master' into span-name-search-support

dbczumar committed:
Add support for searching traces by span.name

- Added span identifier support to SearchTraceUtils
- Implemented span.name filtering in SqlAlchemy store
- Fixed span filter parsing to prevent incorrect tag transformation
- Added explicit join condition for span subqueries
- Added comprehensive tests for span.name search functionality

Supports exact match, LIKE patterns, and NOT EQUAL comparisons.

Signed-off-by: Corey Zumar <zumar.corey@gmail.com>
Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor span filter handling to return separate span_filters

- Modified _get_filter_clauses_for_search_traces to return span_filters separately
- This makes the code cleaner and more explicit about different filter types
- Span filters now have their own explicit join logic separate from tag/metadata

Signed-off-by: Corey Zumar <zumar.corey@gmail.com>
Signed-off-by: dbczumar <corey.zumar@databricks.com>
